### ISSUE TITLE
No reconnection if xfrm interface is enabled

### ISSUE THREAD

**Issue by mastertheknife** [libreswan/libreswan#340]
Hi,
Thank you for this project and your efforts, i really like it, especially the xfrm interfaces, which makes stuff like iptables, tcpdump and dynamic routing, much easier.
However, i am facing some issue, it seems like a bug to me, and i believe it should be reported :-)
Both hosts are running CentOS 8 with libreswan 3.32 (upgraded from 3.31, same issue) from your repository.

Host1:
--------
OS: CentOS 8 with kernel 4.18 (stock, xfrmi support: yes)
Libreswan: 3.31, 3.32
Configuration:
```
conn google-pokimongo
	left=%defaultroute    # Effectively 10.240.0.240
	leftid=Host1IP
	right=Host2IP
	authby=secret
	leftsubnets={10.240.0.0/16,192.168.248.0/24,192.168.246.0/23,192.168.245.0/24,192.168.242.0/23,192.168.241.0/24,10.52.0.0/14}
	rightsubnet=192.168.252.0/24
	ike=aes_gcm256-sha2
	phase2alg=aes_gcm128-null
	auto=start
	encapsulation=no
# Commenting out this line fixes the issue:
	ipsec-interface=1
```

Host2:
--------
OS: CentOS 8 with kernel 4.15 (LXC, host kernel, xfrmi support: no)
Libreswan: 3.31, 3.32
Configuration:
```
conn pokimongo-google
	left=%defaultroute    # Effectively 192.168.252.250
	leftid=Host2IP
	right=Host1IP
	authby=secret
	leftsubnet=192.168.252.0/24
	rightsubnets={10.240.0.0/16,192.168.248.0/24,192.168.246.0/23,192.168.245.0/24,192.168.242.0/23,192.168.241.0/24,10.52.0.0/14}
	ike=aes_gcm256-sha2
	phase2alg=aes_gcm128-null
	auto=start
	encapsulation=no
# xfrmi is not supported in this kernel unfortunately, for now.
	#ipsec-interface=1
```

The issue:
--------------
When the IPSEC tunnel is up, everything is fine.
However, if i restart ipsec on Host2, "systemctl stop ipsec" and then "systemctl start ipsec", or even restart the entire server, the tunnel doesn't come back up. Inside Host1 i see errors regarding bare shunt failed.
Only way to bring the tunnel back up, is to restart ipsec on Host1, this fixes the issue.
If i comment out ipsec-interface on Host1, this issue doesn't exist, i can restart libreswan without issues.

So in summary:
1. ipsec-interface enabled on Host1: Issue happens when restarting libreswan on Host2.
2. ipsec-interface disabled on Host1: Issue doesn't happen when restarting libreswan on Host2.
3. Issue doesn't happen at any case when restarting libreswan on Host1.

Host1 logs:
---------------
[host1_ipsec_status_before.txt](https://github.com/libreswan/libreswan/files/4647336/host1_ipsec_status_before.txt)
[host1_pluto_log_before.txt](https://github.com/libreswan/libreswan/files/4647337/host1_pluto_log_before.txt)
[host1_xfrm_policy_before.txt](https://github.com/libreswan/libreswan/files/4647338/host1_xfrm_policy_before.txt)
[host1_xfrm_state_before.txt](https://github.com/libreswan/libreswan/files/4647339/host1_xfrm_state_before.txt)
[host1_ipsec_status_after.txt](https://github.com/libreswan/libreswan/files/4647340/host1_ipsec_status_after.txt)
[host1_pluto_log_after.txt](https://github.com/libreswan/libreswan/files/4647341/host1_pluto_log_after.txt)
[host1_pluto_log_shutdown_after.txt](https://github.com/libreswan/libreswan/files/4647342/host1_pluto_log_shutdown_after.txt)
[host1_xfrm_policy_after.txt](https://github.com/libreswan/libreswan/files/4647343/host1_xfrm_policy_after.txt)
[host1_xfrm_state_after.txt](https://github.com/libreswan/libreswan/files/4647344/host1_xfrm_state_after.txt)

Host1 working case (ipsec-interface commented out) logs:
-------------------------------------------------------------------------------
[host1_pluto_log_after-noxfrmi.log](https://github.com/libreswan/libreswan/files/4653656/host1_pluto_log_after-noxfrmi.log)

Host2 logs:
---------------
[host2_ipsec_status_before.txt](https://github.com/libreswan/libreswan/files/4647349/host2_ipsec_status_before.txt)
[host2_pluto_log_before.txt](https://github.com/libreswan/libreswan/files/4647350/host2_pluto_log_before.txt)
[host2_xfrm_policy_before.txt](https://github.com/libreswan/libreswan/files/4647351/host2_xfrm_policy_before.txt)
[host2_xfrm_state_before.txt](https://github.com/libreswan/libreswan/files/4647353/host2_xfrm_state_before.txt)
[host2_ipsec_status_after.txt](https://github.com/libreswan/libreswan/files/4647354/host2_ipsec_status_after.txt)
[host2_pluto_log_after.txt](https://github.com/libreswan/libreswan/files/4647355/host2_pluto_log_after.txt)
[host2_xfrm_policy_after.txt](https://github.com/libreswan/libreswan/files/4647356/host2_xfrm_policy_after.txt)
[host2_xfrm_state_after.txt](https://github.com/libreswan/libreswan/files/4647357/host2_xfrm_state_after.txt)

Logs clarification:
Before = Before restarting ipsec (systemctl stop ipsec; small wait; systemctl start ipsec) on Host2
After =  After restarting ipsec (systemctl stop ipsec; small wait; systemctl start ipsec) on Host2

Let me know if you need more information.

Thank you,
Kfir

**Comment by sirajka** [comment]
Hello @mastertheknife I opened issue 341 and saw in your issue you installed 3.32. I was wondering if you had any issues with build?

**Comment by mastertheknife** [comment]
> Hello @mastertheknife I opened issue 341 and saw in your issue you installed 3.32. I was wondering if you had any issues with build?

I'm using their CentOS 8 repository, i did not build these packages.
If you are using CentOS 6, 7 or 8, you can use their repository.

**Comment by sirajka** [comment]
Thanks @mastertheknife I'll try that

**Comment by mastertheknife** [comment]
I just noticed something in the logs that i uploaded, that may be related the issue i'm experiencing.
In Host1 log files "host1_xfrm_policy_after.txt" and "host1_xfrm_state_after.txt", i see some entries with mode "transport".
Where did these come from, i'm using tunnel mode (site2site).

```
src 192.168.246.0/23 dst 192.168.252.0/24 
	dir out priority 2085326 ptype main 
	tmpl src 0.0.0.0 dst 0.0.0.0
		proto esp reqid 0 mode transport
src 192.168.241.0/24 dst 192.168.252.0/24 
	dir out priority 2084814 ptype main 
	tmpl src 0.0.0.0 dst 0.0.0.0
		proto esp reqid 0 mode transport
src 192.168.248.0/24 dst 192.168.252.0/24 
	dir out priority 2084814 ptype main 
	tmpl src 0.0.0.0 dst 0.0.0.0
		proto esp reqid 0 mode transport
src 10.240.0.0/16 dst 192.168.252.0/24 
	dir out priority 2088910 ptype main 
	tmpl src 0.0.0.0 dst 0.0.0.0
		proto esp reqid 0 mode transport
src 192.168.242.0/23 dst 192.168.252.0/24 
	dir out priority 2085326 ptype main 
	tmpl src 0.0.0.0 dst 0.0.0.0
		proto esp reqid 0 mode transport
src 192.168.245.0/24 dst 192.168.252.0/24 
	dir out priority 2084814 ptype main 
	tmpl src 0.0.0.0 dst 0.0.0.0
		proto esp reqid 0 mode transport
src 10.52.0.0/14 dst 192.168.252.0/24 
	dir out priority 2089934 ptype main 
	tmpl src 0.0.0.0 dst 0.0.0.0
		proto esp reqid 0 mode transport
```

```
src 10.240.0.164 dst 192.168.252.169
	proto esp spi 0x00000000 reqid 0 mode transport
	replay-window 0 
	anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
	sel src 10.240.0.164/32 dst 192.168.252.169/32 proto udp sport 1524 dport 5060 dev ipsec1 
src 10.240.0.94 dst 192.168.252.147
	proto esp spi 0x00000000 reqid 0 mode transport
	replay-window 0 
	anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
	sel src 10.240.0.94/32 dst 192.168.252.147/32 proto udp sport 5060 dport 5060 dev ipsec1 
src 10.240.0.13 dst 192.168.252.103
	proto esp spi 0x00000000 reqid 0 mode transport
	replay-window 0 
	anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
	sel src 10.240.0.13/32 dst 192.168.252.103/32 proto udp sport 5060 dport 5060 dev ipsec1 
src 10.240.0.12 dst 192.168.252.103
	proto esp spi 0x00000000 reqid 0 mode transport
	replay-window 0 
	anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
	sel src 10.240.0.12/32 dst 192.168.252.103/32 proto udp sport 5060 dport 5060 dev ipsec1 
```

**Comment by letoams** [comment]
On Tue, 19 May 2020, Kfir Itzhak wrote:

> I just noticed something in the logs that i uploaded, that may be related the issue i'm experiencing.
> In Host1 log files "host1_xfrm_policy_after.txt" and "host1_xfrm_state_after.txt", i see some entries with mode "transport".
> Where did these come from, i'm using tunnel mode (site2site).

If you have auto=route or had packet trigger based connections, those
show up as "transport mode". It's a visual bug mostly AFAIK.

Paul

**Comment by mastertheknife** [comment]
> On Tue, 19 May 2020, Kfir Itzhak wrote: I just noticed something in the logs that i uploaded, that may be related the issue i'm experiencing. In Host1 log files "host1_xfrm_policy_after.txt" and "host1_xfrm_state_after.txt", i see some entries with mode "transport". Where did these come from, i'm using tunnel mode (site2site).
> If you have auto=route or had packet trigger based connections, those show up as "transport mode". It's a visual bug mostly AFAIK. Paul

Both connections are auto=start (should it something different?), and yeah, it seemed packets triggered these reconnects.
Could these existing entries, cause the delete bare shunt failure i'm seeing in the log?
```
May 18 22:31:20 vpn-gateway pluto[2178]: initiate on demand from 10.240.0.240:0 to 192.168.252.250:0 proto=1 because: acquire
May 18 22:31:20 vpn-gateway pluto[2178]: "google-pokimongo/1x0" #55: STATE_V2_CREATE_I: sent IPsec Child req wait response
May 18 22:31:20 vpn-gateway pluto[2178]: "google-pokimongo/1x0" #55: negotiated connection [10.240.0.0-10.240.255.255:0-65535 0] -> [192.168.252.0-192.168.252.255:0-65535 0]
May 18 22:31:20 vpn-gateway pluto[2178]: "google-pokimongo/1x0" #55: STATE_V2_IPSEC_I: IPsec SA established tunnel mode {ESP=>0xaf90d3bc <0xb562dc41 xfrm=AES_GCM_16_128-NONE-MODP2048 NATOA=none NATD=none DPD=passive}
May 18 22:31:23 vpn-gateway pluto[2178]: assign_holdpass() delete_bare_shunt() failed
May 18 22:31:23 vpn-gateway pluto[2178]: initiate_ondemand_body() failed to install negotiation_shunt,
May 18 22:31:23 vpn-gateway pluto[2178]: initiate on demand from 10.240.0.94:5060 to 192.168.252.147:5060 proto=17 because: acquire
May 18 22:31:23 vpn-gateway pluto[2178]: "google-pokimongo/1x0" #56: STATE_V2_CREATE_I: sent IPsec Child req wait response
May 18 22:31:23 vpn-gateway pluto[2178]: "google-pokimongo/1x0" #56: negotiated connection [10.240.0.0-10.240.255.255:0-65535 0] -> [192.168.252.0-192.168.252.255:0-65535 0]
May 18 22:31:23 vpn-gateway pluto[2178]: "google-pokimongo/1x0" #56: STATE_V2_IPSEC_I: IPsec SA established tunnel mode {ESP=>0x1247b9aa <0x59aae31a xfrm=AES_GCM_16_128-NONE-MODP2048 NATOA=none NATD=none DPD=passive}
May 18 22:31:31 vpn-gateway pluto[2178]: assign_holdpass() delete_bare_shunt() failed
May 18 22:31:31 vpn-gateway pluto[2178]: initiate_ondemand_body() failed to install negotiation_shunt,
```
These don't go away with xfrm interface enabled, and i'm typically just rebooting Host1 entirely to bring the tunnel back up.

**Comment by letoams** [comment]
On Tue, 19 May 2020, Kfir Itzhak wrote:

>       On Tue, 19 May 2020, Kfir Itzhak wrote: I just noticed something in the logs that i uploaded, that may be related the issue i'm
>       experiencing. In Host1 log files "host1_xfrm_policy_after.txt" and "host1_xfrm_state_after.txt", i see some entries with mode
>       "transport". Where did these come from, i'm using tunnel mode (site2site).
>       If you have auto=route or had packet trigger based connections, those show up as "transport mode". It's a visual bug mostly AFAIK.
>       Paul
> 
> Both connections are auto=start (should it something different?), and yeah, it seemed packets triggered these reconnects.
> Could these existing entries, cause the delete bare shunt failure i'm seeing in the log?

Yes those are related. What has happened is that auto=start the
connections are supposed to remain up. During rekey, somehow the packets
are lost (because the IKE packet traveled within IPsec?) and the
connection broke. It is putting a %trap or %hold in place to prevent packet leaks
while the tunnel is trying to establish.

> These don't go away with xfrm interface enabled, and i'm typically just rebooting Host1 entirely to bring the tunnel back up.

I think they are caused by the other end using xfrm interface and
triggering the problem?

Paul

**Comment by mastertheknife** [comment]
Only Host1 is using xfrmi currently until the kernel on Host2 is upgraded.
Its very easy to reproduce the issue: Restarting ipsec service on Host2, or rebooting Host2, will trigger it.
The tunnels won't come up again until i stop/start ipsec on Host1, or reboot it.
If xfrmi is disabled, i can't reproduce it. If i restart ipsec on Host2, the tunnels are back within 1-2 seconds.

**Comment by mastertheknife** [comment]
> During rekey, somehow the packets are lost (because the IKE packet traveled within IPsec?) and the connection broke. 

Now that i think about it, it makes sense.
The tunnels use external IPs, however, a firewall is used, and eth0 (the only interface), has an internal IP, which is part of the leftsubnet(s) of each side.
The IP of eth0 in Host1 is 10.240.0.240 and the IP of eth0 in Host2 is 192.168.252.250.

So its possible that IKE tries to send packets over ipsec1, instead of eth0?
I will try to play with left/leftnexthop and report back, thank you!!

**Comment by mastertheknife** [comment]
I tried to put left=10.240.0.240 and leftnexthop=10.240.0.1 (default gw) but that didn't help.
I use external IP for right= on both ends, so i think it doesn't try to do IKE over ipsec1 interface, but i can verify with tcpdump.
Do you need more logs / more debugging? Let me know

**Comment by mastertheknife** [comment]
Attaching log of pluto on Host1, after restarting ipsec on Host2, with xfrmi disabled on Host1 (the working case). Tunnels are up very quickly within 1-2 seconds.

1. At 22:02:19 i ran "systemctl stop ipsec" on Host2 (the other end)
2. At 22:02:29 i ran "systemctl start ipsec" on Host2 (the other end)

[host1_pluto_log_after-noxfrmi.log](https://github.com/libreswan/libreswan/files/4653525/host1_pluto_log_after-noxfrmi.log)

**Comment by cagney** [comment]
Some notes:

- it likely isn't the deleting bare shunt
I suspect it is trying to delete a shunt (kernel policy) that was never inserted.  And same message happens when things work.
Here, 5.0 fixed cases where shunts were installed when they weren't and vis-vis

- it likely isn't blocking IKE
If it were the tunnels wouldn't re-negotiate.

- It is likely, to your point, stray kernel policy
Here 5.0 has two changes: the ipsec-interface is counted so should be deleted when unused; routing and unrouting was improved

I'd wait for 5.0 and see what happens.

**Comment by cagney** [comment]
... or build mainline

**Comment by cagney** [comment]
I'm assuming this is fixed in the soon to be released 5.1, it contains numerous ipsec-interface fixes.
